ktor-http/common/src/io/ktor/http/URLBuilder.kt

```diff
@@ -23,15 +23,15 @@ public const val DEFAULT_PORT: Int = 0
  * @property trailingQuery keep a trailing question character even if there are no query parameters
  */
 public class URLBuilder(
-    public var protocol: URLProtocol = URLProtocol.HTTP,
-    public var host: String = "localhost",
-    public var port: Int = DEFAULT_PORT,
+    public var protocol: URLProtocol? = null,
+    public var host: String? = null,
+    public var port: Int? = null,
     user: String? = null,
     password: String? = null,
-    pathSegments: List<String> = mutableListOf(""),
-    parameters: Parameters = Parameters.Empty,
-    fragment: String = "",
-    public var trailingQuery: Boolean = false
+    pathSegments: List<String>? = null,
```

Comment:
This will make usage of builder harder, but probably there is no better way to do separate default value vs non-set value.
Do you want to proceed with this PR? Otherwise I can take it.

Reply:
@rsinukov You are free to take it :D